test/unit/local/steps/buffer.js

```diff
@@ -0,0 +1,372 @@
+    }
+
+    const push = ({ inputBuffer, inputBatches }) => {
+      const nextBatchNumber = inputBatches.length + 1
+      const newBatch = nonEmptyBatch(nextBatchNumber)
+      inputBatches.push(newBatch)
+      inputBuffer.push(newBatch)
+    }
+
+    const pop = (scenarioState) => {
+      const { outputBuffer, outputBatchesPromise } = scenarioState
+      if (!outputBuffer) {
+        throw new Error(`Step pop cannot occur before asyncMap in scenario`)
+      }
+      scenarioState.outputBatchesPromise = new Promise((resolve, reject) => {
+        outputBatchesPromise
+          .then(outputBatches =>
+            outputBuffer.pop().then(batch =>
+              resolve(outputBatches.concat([batch]))
+            )
+          )
+          .catch(reject)
+      })
+    }
+
+    const scenarioStepFn = step => {
+      switch (step) {
+        case 'map': return map
+        case 'asyncMap': return asyncMap
+        case 'push': return push
+        case 'pop': return pop
+        default: throw new Error(`Unknown scenario step: ${step}`)
+      }
+    }
+
+    /** TRANSFORMS
+     *
+     * The transformation that will be applied to each Batch.
+     *
+     * Must be sync from #map() and async for #asyncMap().
+     */
+
+    const transform = batch => (
+      batch.map(event =>
+        _.defaults(
+          {path: `mapped-${event.path}`},
+          event
+        )
+      )
+    )
+
+    const asyncTransform = async batch => transform(batch)
+
+    /** OPERATIONS */
+
+    describe('#map()', () => {
+      const scenarios = [
+        'map',
+        'map push pop',
+        'map pop push',
+        'map push push pop pop',
+        'map push pop push pop',
+        'map pop push pop push',
+        'map pop push push pop',
+        'push map pop',
+        'push map pop push pop',
+        'push map push pop pop',
+        'push push map pop pop'
+      ]
+
+      /** MOCHA SETUP */
+
+      for (let scenario of scenarios) {
+        describe(scenarioDescription(scenario), () => {
+          let scenarioState /*:: ?: {
+            callback: (Batch) => Batch,
+            inputBatches: Batch[],
+            inputBuffer: Buffer,
+            outputBuffer?: Buffer,
+            outputBatchesPromise: Promise<Batch[]>
+          } */
+
+          beforeEach('init scenarioState', () => {
+            scenarioState = {
+              callback: sinon.spy(transform),
+              inputBatches: [],
+              inputBuffer: new Buffer(),
+              outputBatchesPromise: Promise.resolve([])
+            }
+          })
+
+          for (let step of scenarioSteps(scenario)) {
+            beforeEach(step, () => {
+              // Make sure we don't return a Promise to mocha so scenarios may
+              // reproduce async edge cases.
+              scenarioStepFn(step)(scenarioState)
+            })
+          }
+
+          beforeEach('await output', () => scenarioState.outputBatchesPromise)
```

Comment:
I don't quite understand what we're waiting here and why.

Reply:
Although `Buffer#map()` applies a synchronous transform, `#pop()` is still async.
And the `pop` steps are non-blocking, so we may test non-waiting push/pop combos.

So at the end, unless we await for all the `#pop()` to complete (using this promise), all the callback assertions will fail.
The new output Buffer assertion doesn't care and the output batches one will work since it's explicitely awaiting for the same promise to get the batches.

It seemed logical to me to await before assertions, otherwise the scenario run is not yet complete in some way...

Reply:
Would an improved hook description help? E.g. `wait for output batches to be available`?

Reply:
No it's alright. I just missed the fact that all the actions are run in a `beforeEach` callback prior to this one.

I get it now :).